File: lexicon-android/app/src/test/java/com/robertvokac/lexicon/ui/graph/GraphLayoutTest.kt
package com.robertvokac.lexicon.ui.graph

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.hypot

class GraphLayoutTest {
    @Test
    fun itemsStayApartTheCentreStaysAndTheLayoutRepeats() {
        val depths = listOf(0, 1, 1, 1, 2, 2, 2, 2, 2)
        val edges = listOf(0 to 1, 0 to 2, 0 to 3, 1 to 4, 1 to 5, 2 to 6, 3 to 7, 3 to 8)
        val points = GraphLayout.layout(depths, edges)
        assertEquals(GraphLayout.Point(0.0, 0.0), points[0])
        var closest = Double.MAX_VALUE
        for (i in points.indices) for (j in i + 1 until points.size) {
            closest = minOf(closest, hypot(points[i].x - points[j].x, points[i].y - points[j].y))
        }
        assertTrue("closest $closest", closest > 60)
        assertEquals(points, GraphLayout.layout(depths, edges))
    }
}

File: lexicon-web/tests/graphlayout.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import { layoutGraph } from '../js/graphlayout.js';

test('the layout keeps items apart, the centre fixed, and is repeatable', () => {
    const depths = [0, 1, 1, 1, 2, 2, 2, 2, 2];
    const edges = [[0, 1], [0, 2], [0, 3], [1, 4], [1, 5], [2, 6], [3, 7], [3, 8]];
    const points = layoutGraph(depths, edges);
    assert.deepEqual(points[0], { x: 0, y: 0 });
    let closest = Infinity;
    for (let i = 0; i < points.length; i += 1) {
        for (let j = i + 1; j < points.length; j += 1) {
            closest = Math.min(closest, Math.hypot(points[i].x - points[j].x, points[i].y - points[j].y));
        }
    }
    assert.ok(closest > 60, `closest ${closest}`);
    assert.deepEqual(layoutGraph(depths, edges), points);
});
